### PR TITLE
Enhance circular buffer

### LINKED ISSUE #276: Additional unit tests for Circular Buffer
**What has to be done**
https://github.com/RespiraWorks/VentilatorSoftware/pull/271 introduces a circular buffer library and some unit tests.

We should add some more unit tests, as described at the bottom of https://github.com/RespiraWorks/VentilatorSoftware/blob/master/controller/test/circ_buff/circ_buff_test.cpp

**How do you know it has to be done**
A circular buffer is a pretty nasty place to have a bug, yet easy enough to test exhaustively - so this is what we must do.

**Code pointers**
See above.

### PR DESCRIPTION
**Pull request by asmodai27**
Make the template safe against N<0 and N=0
Enhance existing tests and add tests
Closes #276

### CODE REVIEW

File: software/controller/test/circ_buff/circ_buff_test.cpp [hunk 1]
```diff
@@ -18,14 +18,12 @@ limitations under the License.
 #include <cstdlib>
 #include <optional>
 
-// Just getting my feet wet with gtest
+// Test the buffer counts and
```
Comment (martukas) [inline software/controller/test/circ_buff/circ_buff_test.cpp:21]: This comment looks incomplete?

File: software/controller/lib/core/circular_buffer.h [hunk 2]
```diff
@@ -66,23 +69,23 @@ template <class T, int N> class CircBuff {
   // Returns false if the buffer is full.
   [[nodiscard]] bool Put(T dat) {
     BlockInterrupts block;
-    int h = head + 1;
-    if (h >= N) {
+    int h = head_ + 1;
```
Comment (martukas) [inline software/controller/lib/core/circular_buffer.h:72]: is `h` maybe `new_head`?

File: software/controller/lib/core/circular_buffer.h [hunk 3]
```diff
@@ -26,37 +26,40 @@ limitations under the License.
...
+  volatile T buff_[N + 1];
+  volatile int head_, tail_;
 
 public:
-  CircBuff() { head = tail = 0; }
+  CircBuff() { head_ = tail_ = 0; }
 
   // Return number of elements available in the buffer to read.
-  int FullCt() {
+  int FullCt() const {
```
Comment (martukas) [inline software/controller/lib/core/circular_buffer.h:40]: Maybe these functions as well at the class can use full words? Like "count" and "circular buffer"?

**Review by martukas — APPROVED** [review]
Sorry this took a while.
I think functionally this is good, but made some comments about readability/style.

And curious: are we possibly reinventing the bike? Maybe there is already a well maintained library that does this that could be ported? I seem to recall people implementing circular buffers at my previous job...

File: software/controller/test/debug/trace_test.cpp [hunk 4]
```diff
@@ -104,7 +104,7 @@ TEST(Trace, BufferFull) {
   trace.SetFlags(1);
 
   // A buffer of size N actually holds only N-1 values.
```
Comment (asmodai27) [inline software/controller/test/debug/trace_test.cpp:106]: update comment as well

**Review by asmodai27 — COMMENTED** [review]
There are probably other options but it is simple and straightforward enough that we don't need to look into those, now that this is integrated.
I will address the style comments before merging.

**Pull request merged**